src/plots/cartesian/dragbox.js

```diff
@@ -178,7 +183,19 @@ function makeDragBox(gd, plotinfo, x, y, w, h, ns, ew) {
             prepSelect(e, startX, startY, dragOptions, dragModeNow);
         } else {
             dragOptions.clickFn = clickFn;
-            clearAndResetSelect();
+            if(isSelectOrLasso(dragModePrev)) {
+                // TODO Fix potential bug
+                // Note: clearing / resetting selection state only happens, when user
+                // triggers at least one interaction in pan/zoom mode. Otherwise, the
+                // select/lasso outlines are deleted (in plots.js.cleanPlot) but the selection
+                // cache isn't cleared. So when the user switches back to select/lasso and
+                // 'adds to a selection' with Shift, the "old", seemingly removed outlines
+                // are redrawn again because the selection cache still holds their coordinates.
+                // However, this isn't easily solved, since plots.js would need
+                // to have a reference to the dragOptions object (which holds the
+                // selection cache).
+                clearAndResetSelect();
```

Comment:
Here's a gif that hopefully clearifies things (as we might need input from @alexcjohnson )

![peek 2018-08-29 12-18](https://user-images.githubusercontent.com/6675409/44800973-ca6de580-ab85-11e8-9b85-3d65b334fa99.gif)

Note that the go from select -> pan -> select w/o actually panning. 

Personally, I wouldn't consider this a bug as the "cached" selection outlines are still "correct" (i.e. they are drawn at the correct x/y coordinates). Now, is this the "ideal" behavior, I'm not so sure.

Reply:
I like the behavior that we don't clear the selection itself until the last possible moment... but it would certainly be nice if the outline were redrawn when you switch back to a select mode. It's a sufficiently uncommon edge case though that I'd rather not hold up the PR for it - just leave it as is for now and make a new issue to follow up on this later.

Reply:
I created issue #2955 referencing this discussion here.